astropy/table/serialize.py

```diff
@@ -51,52 +56,44 @@ def _represent_mixin_as_column(col, name, new_cols, mixin_cols,
     # - description: DO store
     # - meta: DO store
     info = {}
-    for attr, nontrivial, xform in (('unit', lambda x: x not in (None, ''), str),
+    for attr, nontrivial, xform in (('unit', lambda x: x is not None and x != '', str),
                                     ('format', lambda x: x is not None, None),
                                     ('description', lambda x: x is not None, None),
                                     ('meta', lambda x: x, None)):
         col_attr = getattr(col.info, attr)
         if nontrivial(col_attr):
             info[attr] = xform(col_attr) if xform else col_attr
 
-    obj_attrs = col.info._represent_as_dict()
-    ordered_keys = col.info._represent_as_dict_attrs
-
     data_attrs = [key for key in ordered_keys if key in obj_attrs and
                   getattr(obj_attrs[key], 'shape', ())[:1] == col.shape[:1]]
 
     for data_attr in data_attrs:
         data = obj_attrs[data_attr]
-        if len(data_attrs) == 1 and not has_info_class(data, MixinInfo):
```

Comment:
Replace this ad-hoc rule (and mostly repeated code) with explicit test that uses `col.info._represent_as_dict_primary_data`.

Reply:
Much better indeed (even if not completely sold on profusion of names in `info` - at some point, we should have a nice, explicit API, but right now I think we're still finding out what needs to be in it!).

p.s. Having looked further down now: Do very much like the added comments.